File: desktop/src/main/services/device-service.ts
/**
 * Device registration and authentication service.
 *
 * Coordinates:
 * - API client for backend communication
 * - Secure storage for credential persistence
 * - Device model for type safety
 */

import { apiClient, ApiError, ApiResult } from '../api/client';
import { secureStorage } from './secure-storage';
import { config } from '../config';
import {
  DeviceCredentials,
  DeviceInfo,
  DeviceRegistrationResponse,
  DeviceType,
  toDeviceCredentials,
  toDeviceInfo,
} from '../models/device';

export enum AuthState {
  UNINITIALIZED = 'uninitialized',
  REGISTERING = 'registering',
  AUTHENTICATED = 'authenticated',
  ERROR = 'error',
}

export interface DeviceState {
  authState: AuthState;
  deviceInfo: DeviceInfo | null;
  error: string | null;
}

/** Device service for managing registration and auth state. */
export class DeviceService {
  private state: DeviceState = {
    authState: AuthState.UNINITIALIZED,
    deviceInfo: null,
    error: null,
  };

  private stateListeners: Array<(state: DeviceState) => void> = [];

  /**
   * Get current authentication state.
   */
  getState(): DeviceState {
    return { ...this.state };
  }

  /**
   * Subscribe to state changes.
   */
  subscribe(listener: (state: DeviceState) => void): () => void {
    this.stateListeners.push(listener);
    return () => {
      const idx = this.stateListeners.indexOf(listener);
      if (idx >= 0) this.stateListeners.splice(idx, 1);
    };
  }

  private setState(partial: Partial<DeviceState>): void {
    this.state = { ...this.state, ...partial };
    this.notifyListeners();
  }

  private notifyListeners(): void {
    for (const listener of this.stateListeners) {
      listener(this.getState());
    }
  }

  /**
   * Initialize the device service.
   * Checks for existing credentials; if none, registers the device.
   */
  async initialize(): Promise<void> {
    // Check for existing credentials
    const existing = await this.loadCredentials();
    // If loadCredentials failed, it already set ERROR state - don't proceed to registration
    if (this.state.authState === AuthState.ERROR) {
      return;
    }
    if (existing) {
      this.setState({
        authState: AuthState.AUTHENTICATED,
        deviceInfo: toDeviceInfo(existing),
        error: null,
      });
      return;
    }

    // No credentials - register
    await this.registerDevice();
  }

  /**
   * Load credentials from secure storage.
   */
  async loadCredentials(): Promise<DeviceCredentials | null> {
    try {
      return await secureStorage.getCredentials();
    } catch (err) {
      const message = err instanceof Error ? err.message : 'Unknown storage error';
      this.setState({
        authState: AuthState.ERROR,
        error: `Failed to load credentials: ${message}`,
      });
      return null;
    }
  }

  /**
   * Register this device with the backend.
   */
  async registerDevice(): Promise<DeviceCredentials | null> {
    this.setState({
      authState: AuthState.REGISTERING,
      error: null,
    });

    try {
      const result = await apiClient.registerDevice(config.deviceName, DeviceType.DESKTOP);

      if (!result.ok) {
        this.setState({
          authState: AuthState.ERROR,
          error: `Registration failed: ${result.error.message}`,
        });
        return null;
      }

      const credentials = toDeviceCredentials(result.data);

      // Store securely
      try {
        await secureStorage.storeCredentials(credentials);
      } catch (err) {
        const message = err instanceof Error ? err.message : 'Unknown storage error';
        this.setState({
          authState: AuthState.ERROR,
          error: `Failed to store credentials: ${message}`,
        });
        return null;
      }

      this.setState({
        authState: AuthState.AUTHENTICATED,
        deviceInfo: toDeviceInfo(credentials),
        error: null,
      });

      return credentials;
    } catch (err) {
      const message = err instanceof Error ? err.message : 'Unknown error';
      this.setState({
        authState: AuthState.ERROR,
        error: `Registration error: ${message}`,
      });
      return null;
    }
  }

  /**
   * Get stored auth token for authenticated requests.
   */
  async getAuthToken(): Promise<string | null> {
    const credentials = await this.loadCredentials();
    return credentials?.authToken ?? null;
  }

  /**
   * Get device info (for UI).
   */
  getDeviceInfo(): DeviceInfo | null {
    return this.state.deviceInfo;
  }

  /**
   * Clear credentials and reset state (for testing/logout).
   */
  async clearCredentials(): Promise<void> {
    try {
      await secureStorage.deleteCredentials();
    } catch {
      // Ignore errors on delete
    }
    this.setState({
      authState: AuthState.UNINITIALIZED,
      deviceInfo: null,
      error: null,
    });
  }
}

/** Default device service instance. */
export const deviceService = new DeviceService();

File: mobile/lib/services/device_service.dart
// ignore_for_file: dangling_library_doc_comments, unnecessary_library_name, prefer_initializing_formals

/// Device service for Synk mobile.
///
/// Handles device registration, device info retrieval,
/// and secure storage of device credentials.

import 'package:synk_mobile/core/config/app_config.dart';
import 'package:synk_mobile/core/network/api_client.dart';
import 'package:synk_mobile/core/storage/secure_storage.dart';
import 'package:synk_mobile/models/device.dart';

/// Service for device-related operations.
class DeviceService {
  final ApiClient _apiClient;

  /// Creates a device service with the given API client.
  DeviceService({required ApiClient apiClient}) : _apiClient = apiClient;

  /// Registers a new device with the backend.
  ///
  /// [deviceName] - Human-readable name for the device.
  /// [deviceType] - Type of device (mobile/desktop).
  ///
  /// Returns the device registration response containing device_id,
  /// device_name, device_type, and auth_token.
  ///
  /// Throws [ApiException] on failure.
  Future<DeviceRegistrationResponse> registerDevice({
    required String deviceName,
    required DeviceType deviceType,
  }) async {
    final request = DeviceRegistrationRequest(
      deviceName: deviceName.trim(),
      deviceType: deviceType,
    );

    final response = await _apiClient.post<DeviceRegistrationResponse>(
      '/devices/register',
      body: request.toJson(),
      requiresAuth: false, // Registration doesn't require auth
      parser: (json) => DeviceRegistrationResponse.fromJson(json as Map<String, dynamic>),
    );

    // Store credentials securely after successful registration
    await _storeCredentials(response);

    // Set auth token on API client for future requests
    _apiClient.setAuthToken(response.authToken);

    return response;
  }

  /// Retrieves the current device information from the backend.
  ///
  /// Requires authentication token to be set on the API client.
  ///
  /// Returns the device information.
  ///
  /// Throws [ApiException] on failure (e.g., 401 if token invalid).
  Future<Device> getCurrentDevice() async {
    final deviceId = await getStoredDeviceId();
    if (deviceId == null) {
      throw const ApiException(
        statusCode: 401,
        message: 'No device registered. Please register first.',
      );
    }

    final response = await _apiClient.get<Device>(
      '/devices/$deviceId',
      requiresAuth: true,
      parser: (json) => Device.fromJson(json as Map<String, dynamic>),
    );

    return response;
  }

  /// Checks if a device is already registered locally.
  Future<bool> isRegistered() async {
    return await SecureStorage.containsKey(key: AppConfig.storageKeyDeviceId);
  }

  /// Gets the stored device ID.
  Future<String?> getStoredDeviceId() async {
    return await SecureStorage.read(key: AppConfig.storageKeyDeviceId);
  }

  /// Gets the stored auth token.
  Future<String?> getStoredAuthToken() async {
    return await SecureStorage.read(key: AppConfig.storageKeyAuthToken);
  }

  /// Gets the stored device name.
  Future<String?> getStoredDeviceName() async {
    return await SecureStorage.read(key: AppConfig.storageKeyDeviceName);
  }

  /// Loads stored credentials and configures the API client.
  ///
  /// Call this on app startup to restore authenticated state.
  Future<void> loadStoredCredentials() async {
    final deviceId = await getStoredDeviceId();
    final authToken = await getStoredAuthToken();

    if (deviceId != null && authToken != null) {
      _apiClient.setAuthToken(authToken);
    }
  }

  /// Stores device credentials securely.
  Future<void> _storeCredentials(DeviceRegistrationResponse response) async {
    await Future.wait([
      SecureStorage.write(
        key: AppConfig.storageKeyDeviceId,
        value: response.deviceId,
      ),
      SecureStorage.write(
        key: AppConfig.storageKeyAuthToken,
        value: response.authToken,
      ),
      SecureStorage.write(
        key: AppConfig.storageKeyDeviceName,
        value: response.deviceName,
      ),
    ]);
  }

  /// Clears all stored credentials and resets the API client.
  ///
  /// Use this for logout or device unregistration.
  Future<void> clearCredentials() async {
    await SecureStorage.deleteAll();
    _apiClient.clearAuthToken();
  }
}